### PR TITLE
The Dangers of Teleportation

### PR DESCRIPTION
**Pull request by LordFowl**
Teleportation using the do_teleport() proc will now have violent effects if you teleport into an unoccupied space. This currently does not effect teleportation that does not use the do_teleport() proc (wizards, ninjas, cultists).

If there is an object in your destination, and that object is both anchored and dense or opaque (turfs, tables, etcetera), then you will partially phase into the object. If you are a mob you will lose a limb, and if you are an object you will become damaged (if the object has an ex_act() defined). Then an explosion will occur based on your size (Most explosions are large enough only to damage the destination turf itself). You will be shunted into the nearest unoccupied space.

If there is NO adjacent unoccupied space, you will instead be destroyed. Mobs will gib, and objects will have crush_act called. The resulting explosion will be larger.

image

image

image

image

### CODE REVIEW

**Comment by Kaedwuff** [comment]
Won't this make telescience prohibitively dangerous to play with?

**Comment by LordFowl** [comment]
No.

File: code/datums/helper_datums/teleport.dm [hunk 1]
```diff
@@ -113,9 +113,76 @@
...
+						L.forceMove(newdest)
+					else
+						to_chat(L, "<span class='danger'>Your life flashes before your eyes as you phase into \the [impediment] before the universe suddenly and violently corrects itself!</span>")
+						L.gib()
+
+			if(!newdest)
+				boominess += 10
+
+			destturf.visible_message("<span class ='danger'>There is a sizable emission of energy as \the [teleatom] phases into \the [impediment]!</span>")
+			explosion(destturf, ((boominess > 10) ? 1 : 0), ((boominess > 5) ? (boominess/5) : 0), boominess/5, boominess/2)
```
Comment (Sindorman) [inline code/datums/helper_datums/teleport.dm]: I am entirely against the idea of any sort of explosion, unless we teleport volatile things into dense object. Even on that case explosion is very low chance.

According to how teleportation works - you disassemble atoms in one place and assemble into another place, the collision of atoms results into what Github calls merge conflict. Result of the conflict makes two atoms merge into one atom creating a hybrid of both object. The explosion should never occur unless two object contained atoms that in combination create combustion and tiny explosion(by tiny we are talking atom size explosion).

Comment (LordFowl) [reply code/datums/helper_datums/teleport.dm]: I disagree.

Comment (Sindorman) [reply code/datums/helper_datums/teleport.dm]: What do you mean desire? First we do not have lore about teleportation, because I checked the wiki page for it and nothing mentions how it works. Everything is always explained as bluespace. If the time is involved into teleportation then I can agree that explosions might occur. But the only way it can at that point is if at certain point of time this atom had something that would create combustion with current atom. If we take into account the alternate dimensions that could justify it - because in some other universe on place of that wall can be fuel tank or something else. But again given that this theory leads us that alternate dimensions are infinite - that means that chances of having explosion are close to 0.

To summarize: with your justification of why it happens you need to make explosion a low probability thing.

Comment (LordFowl) [reply code/datums/helper_datums/teleport.dm]: No.

Comment (Sindorman) [reply code/datums/helper_datums/teleport.dm]: What do you mean no?

File: code/datums/helper_datums/teleport.dm [hunk 2]
```diff
@@ -113,9 +113,76 @@
...
+					boominess += 10
+
+			if(istype(teleatom, /mob/living))
+				var/mob/living/L = teleatom
+				boominess += L.mob_size/4
+				if(istype(L, /mob/living/carbon/human))
+					var/mob/living/carbon/human/H = L
+					if(newdest)
+						var/obj/item/organ/external/E = pick(H.organs)
+						to_chat(H, "<span class='danger'>You partially phase into \the [impediment], causing your \the [E] to bloodily dematerialize!</span>")
```
Comment (Arrow768) [inline code/datums/helper_datums/teleport.dm]: I dont think the second \the makes sense.
`causing your the right arm to bloodily dematerialize!`

File: code/datums/helper_datums/teleport.dm (hunk 1 above)
Comment (LordFowl) [reply code/datums/helper_datums/teleport.dm]: I mean that I'm not adding in a "probability" of explosions, I do not find your justifications convincing, and further I don't really see a need to justify the explosive nature of energy emission any further than I already have. The brass tacks can be divvied up by someone who cares about that sort of thing.

**Comment by Kaedwuff** [comment]
Both Wizards and ninjas have code built into their teleports that prevent them from going into solid objects. 

Unless there is an interaction between these two bits of codes that make them hurt themselves even as they're pushed out of the wall then it won't be a problem

**Comment by LordFowl** [comment]
More importantly, as the PR itself states _in the description_ "This currently does not effect teleportation that does not use the do_teleport() proc (wizards, ninjas, cultists)."

**Comment by Butterrobber202** [comment]
> More importantly, as the PR itself states _in the description_ "This currently does not effect teleportation that does not use the do_teleport() proc (wizards, ninjas, cultists)."

Thank, I’m a blind meme

**Comment by MarinaGryphon** [comment]
I think this will just make telescience able to be used as a weapon. Teleport a crate into occupied space near the AI core, for example (which there is plenty of thanks to double or triple-thick walls) and you've got yourself an easy, uncounterable bomb. Repeat ad nauseum.

**Comment by Sindorman** [comment]
@MarinaGryphon That is why we have PR #5652 which will go well with this PR - protecting areas like Vault  and AI core, maybe bridge, etc.
